Add rpmsg-echo-test example

TanmayShah-xilinx opened the pull request:
rpmsg-echo-test example demonstrates how to write linux userspace application that uses rpmsg devices in kernel space. This application sends pre-defined payload to remote processor firmware and waits for remote processor to send same payload back. It then verifies received payload integrity. Number of packets to send to remote processor can be configured from command line

Signed-off-by: Tanmay Shah <tanmay.shah@amd.com>

arnopo commented on examples/linux/rpmsg-echo-test/echo_test.c:
```diff
@@ -0,0 +1,330 @@
+ * context. The application sends chunks of data to the
+ * remote processor. The remote side echoes the data back
+ * to application which then validates the data returned.
+ */
```
An README.txt would be nice to explain the test and what is needed on remote side ( compatible plaftform, associated remoteproc project...)

TanmayShah-xilinx replied:
Thanks. Yes that makes sense. I will add such README.

TanmayShah-xilinx commented on Makefile:
```diff
@@ -0,0 +1,10 @@
+TARGETS := all clean
```
Ok I will remove it. I agree for now we don't have to push generic makefile yet.

TanmayShah-xilinx replied:
Yes. This is old demo where by default dst buffer was picked up by driver from memory-region (vring). We have some new patch that allows to set destination buffer from command line. I will create new pull request accordingly.

TanmayShah-xilinx commented on examples/linux/rpmsg-echo-test/echo_test.c:
```diff
@@ -0,0 +1,330 @@
+	eptinfo.src = 0;
+	eptinfo.dst = 0xFFFFFFFF;
+	ret = rpmsg_create_ept(charfd, &eptinfo);
+	if (ret) {
```
Yes. This is old demo where for 0xFFFFFFFF  dst address, by default driver picked up from memory-region (vrings). We have some new patch that allows to set destination buffer from command line which sets it in sysfs. I will create new pull request that has latest source code as well.

arnopo replied:
Is there a reason to create a specific RP for these patches? Perhaps you could merge them in the code or add these patches into this PR.
Since this code should become a reference code for the use of rpmsg_char for Linux, I would expect it to be compatible with the latest version of Linux.

TanmayShah-xilinx replied:
Yes I will squash all the patches in same PR. Also, demos are compatible with latest version of linux.

TanmayShah-xilinx commented:
I will close this pull request and open new one as new echo_test app contains more changes.

TanmayShah-xilinx commented:
As discussed this with Arnaud same pull request can be used.

TanmayShah-xilinx commented:
@arnopo 

I addressed comments in latest commit.
- No makefile in root dir
- Add README for echo_test
- Allow end point dst address to be set from command line.

edmooring reviewed (approved):
Looks good to me.

arnopo commented on examples/linux/rpmsg-echo-test/echo_test.c:
```diff
@@ -0,0 +1,413 @@
+		perror("write SHUTDOWN_MSG\n");
+}
+
+int rpmsg_create_ept(int rpfd, struct rpmsg_endpoint_info *eptinfo)
```
Please rename this function, that introduce confusing as already defined in the open-amp library

TanmayShah-xilinx commented:
@wmamills, @arnopo, @edmooring 

I have split changes in different commits as requested. If all look good, I will do the same for other examples as well.

arnopo commented:
> @wmamills, @arnopo, @edmooring
> 
> I have split changes in different commits as requested. If all look good, I will do the same for other examples as well.

look good to me
Thanks for the extra work

wmamills reviewed (changes requested):
Reference should be to the openamp lib example not the Xilinx sample

TanmayShah-xilinx commented:
> Reference should be to the openamp lib example not the Xilinx sample

Ah! I understand now. That would be: https://github.com/OpenAMP/open-amp/blob/main/apps/examples/echo/rpmsg-echo.c

I will update README accordingly in all the examples.

TanmayShah-xilinx commented:
> > Reference should be to the openamp lib example not the Xilinx sample
> 
> Ah! I understand now. That would be: https://github.com/OpenAMP/open-amp/blob/main/apps/examples/echo/rpmsg-echo.c
> 
> I will update README accordingly in all the examples.

Done.

TanmayShah-xilinx commented:
@arnopo , @edmooring 
Could you please look at latest change ?
I can only re-request review for one of you.
I am not sure why I can't re-request reviews for both of you.

TanmayShah-xilinx commented on examples/linux/rpmsg-echo-test/echo_test.c:
```diff
@@ -306,9 +307,15 @@ int main(int argc, char *argv[])
-		return charfd;
+
+	/* kernel >= 6.0 has new path for rpmsg_ctrl device */
+	charfd = get_rpmsg_chrdev_fd(rpmsg_ctrl_dev_name, rpmsg_char_name);
```
Yes correct, I will add modprobe(rpmsg_ctrl.ko) in system command. 
When I tested with 6.0, built rpmsg_* drivers as part of kernel so I didn't see that issue.

Thanks.

arnopo reviewed (approved):
LGTM
